master/internal/cache/cache.go

```diff
@@ -0,0 +1,199 @@
+)
+
+type modelDefFolder struct {
+	fileTree   []*experimentv1.FileNode
+	cachedTime time.Time
+	lock       sync.RWMutex
+	path       string
+}
+
+// FileCache is metadata for files cached at file system.
+type FileCache struct {
+	rootDir string
+	maxAge  time.Duration
+	caches  map[int]*modelDefFolder
+	lock    sync.Mutex
+}
+
+// NewFileCache initialize FileCache obj.
+func NewFileCache(rootDir string, maxAge time.Duration) *FileCache {
+	err := os.RemoveAll(rootDir)
+	if err != nil {
+		log.WithError(err).Errorf("failed to clear the content of cache folder at %s", rootDir)
+	}
+	return &FileCache{
+		rootDir: rootDir,
+		maxAge:  maxAge,
+		caches:  make(map[int]*modelDefFolder),
+	}
+}
+
+func (f *FileCache) getOrCreateFolder(expID int) (*modelDefFolder, error) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	value, ok := f.caches[expID]
+	if ok {
+		return value, nil
+	}
+
+	exp := struct {
+		ModelDefinition []byte
+	}{}
+	err := db.Bun().NewSelect().TableExpr(
+		"experiments").Column("model_definition").Where("id = ?", expID).Scan(context.Background(), &exp)
+	if err != nil {
+		return nil, err
+	}
+	var fileTree []*experimentv1.FileNode
+	arc, err := archive.FromTarGz(exp.ModelDefinition)
+	if err != nil {
+		return nil, err
+	}
+	err = os.MkdirAll(f.genPath(expID, ""), 0700)
+	if err != nil {
+		return nil, err
+	}
+	for _, ar := range arc {
+		if ar.IsDir() {
+			err = os.MkdirAll(f.genPath(expID, ar.Path), 0700)
+		} else {
+			err = os.WriteFile(f.genPath(expID, ar.Path), ar.Content, 0600)
+		}
+		if err != nil {
+			return nil, err
+		}
+		fileTree = append(fileTree, &experimentv1.FileNode{
+			Path:          ar.Path,
+			ModifiedTime:  timestamppb.New(ar.ModifiedTime.Time),
+			ContentLength: int32(len(ar.Content)),
+			IsDir:         ar.IsDir(),
+			ContentType:   http.DetectContentType(ar.Content),
+			Name:          filepath.Base(ar.Path),
+		})
+	}
+	value = &modelDefFolder{
+		path:       f.genPath(expID, ""),
+		fileTree:   fileTree,
+		cachedTime: time.Now(),
+	}
+	f.caches[expID] = value
+	f.prune()
+
+	return value, nil
+}
+
+func (f *FileCache) prune() {
+	for expID, folder := range f.caches {
+		if folder.cachedTime.Add(f.maxAge).Before(time.Now()) {
+			err := os.RemoveAll(folder.path)
+			if err != nil {
+				log.WithError(err).Errorf("failed to prune model definition cache under %s", folder.path)
+			}
+			delete(f.caches, expID)
+		}
+	}
+}
+
+func (f *FileCache) genPath(expID int, path string) string {
+	return filepath.Join(f.rootDir, strconv.Itoa(expID), path)
+}
+
+func (f *FileCache) getFileTree(expID int) ([]*experimentv1.FileNode, *modelDefFolder, error) {
+	folder, err := f.getOrCreateFolder(expID)
+	if err != nil {
+		return nil, nil, err
+	}
+	folder.lock.RLock()
+	defer folder.lock.RUnlock()
+	return folder.fileTree, folder, nil
+}
+
+// GetFileTreeNested returns folder tree structure with given experiment id.
+func (f *FileCache) GetFileTreeNested(expID int) ([]*experimentv1.FileNode, error) {
```

Comment:
This is just a nit, but I'm pretty sure that in a go interface, you're [not supposed](https://go.dev/doc/effective_go#Getters) to include the word "get" in a getter's name.  So it might be worth considering removing this.

Reply:
Sure, thank you for sharing, this has been fixed